torch/optim/adagrad.py

```diff
@@ -63,6 +64,7 @@ def step(self, closure=None):
                 clr = group['lr'] / (1 + (state['step'] - 1) * group['lr_decay'])
 
                 if p.grad.data.is_sparse:
+                    grad.contiguous()  # the update is non-linear so indices must be unique
```

Comment:
I find it a bit confusing that for sparse tensors, `.contiguous()` operates in-place, contrary to dense tensors. But I suppose sparse tensors are meant for internal use only?